recirq/qcqmc/blueprint.py

```diff
@@ -0,0 +1,318 @@
+# Copyright 2024 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+from typing import Dict, Iterable, Iterator, List, Optional, Sequence, Tuple, Union
+
+import attrs
+import cirq
+import numpy as np
+import quaff
+
+from recirq.qcqmc import config, data, trial_wf, for_refactor
+
+BlueprintParams = Union["BlueprintParamsTrialWf", "BlueprintParamsRobustShadow"]
+
+
+def apply_optimizer_suite_0(circuit: cirq.Circuit) -> cirq.Circuit:
+    """A circuit optimization routine that tries to merge gates.
+
+    Args:
+        circuit: The circuit to optimize
+
+    Returns:
+        The gate optimized circuit.
+    """
+
+    circuit = cirq.expand_composite(circuit)
+    circuit = cirq.align_left(circuit)
+
+    circuit = cirq.drop_empty_moments(circuit)
+
+    return circuit
+
+
+def _to_tuple_of_tuples(
+    x: Iterable[Iterable[cirq.Qid]],
+) -> Tuple[Tuple[cirq.Qid, ...], ...]:
+    return tuple(tuple(_) for _ in x)
+
+
+def _to_tuple(x: Iterable[cirq.Circuit]) -> Tuple[cirq.Circuit, ...]:
```

Comment:
This seems extraneous.  Why not just use tuple()?

Reply:
I think mypy complains about tuple as the converter similar to https://github.com/python/mypy/issues/8389 (not that it's really a problem since we're not type checking but my editor yells at me.

Reply:
type annotations in a dataclass probably

Reply:
Oh, okay, let's put a comment in that this is needed for type conversion in the dataclass then.

Reply:
Done.